packages/components/src/navigation/children-utils.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * External dependencies
+ */
+import { keyBy, omit } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { ROOT_MENU } from './constants';
+import NavigationItem from './item';
+import NavigationMenu from './menu';
+
+/**
+ *
+ * @typedef {Object} ChildWithClosestParent
+ * @property {WPElement} child element that matched the predicate
+ * @property {WPElement?} parent closest matching parent that matched the parentPredicate
+ */
+
+/**
+ *
+ * @param {WPElement} initialChildren children to iterate through recursively
+ * @param {Function} predicate children matching this function predicate will be returned
+ * @param {Function} parentPredicate closest parent will be determined by this function predicate
+ *
+ * @return {ChildWithClosestParent[]} children with their closest parent
+ */
+export const findChildrenWithClosestParent = (
+	initialChildren,
+	predicate,
+	parentPredicate
+) => {
+	const hasPredicate = typeof predicate === 'function';
+	const hasParentPredicate = typeof parentPredicate === 'function';
+
+	const innerRecursion = ( children, lastParent ) => {
+		let items = [];
+
+		if ( ! Array.isArray( children ) ) {
+			children = [ children ];
+		}
+
+		for ( const child of children ) {
+			if ( ! child ) {
+				continue;
+			}
+
+			const predicateOk = hasPredicate && predicate( child );
+			if ( predicateOk ) {
+				items.push( { child, parent: lastParent } );
+			}
+
+			const hasChildren = child?.props?.children;
+			if ( hasChildren ) {
+				const parentPredicateOk =
+					hasParentPredicate && parentPredicate( child );
+				if ( parentPredicateOk ) {
+					lastParent = child;
+				}
+
+				items = items.concat(
+					findChildrenWithClosestParent(
+						child.props.children,
+						predicate,
+						parentPredicate,
+						lastParent
+					)
+				);
+			}
+		}
+
+		return items;
+	};
+
+	return innerRecursion( initialChildren, null );
+};
+
+export const findNavigationItems = ( children ) => {
+	const items = findChildrenWithClosestParent(
+		children,
+		( { type } ) => type === NavigationItem
+	);
+
+	const itemsWithMenu = items.map( ( { child, parent } ) => ( {
+		...child,
+		props: { ...child.props, menu: parent?.props?.menu ?? ROOT_MENU },
+	} ) );
+
+	const itemsWithoutChildren = itemsWithMenu.map( ( { props } ) =>
+		omit( props, 'children' )
+	);
+	const itemsByItemProp = keyBy( itemsWithoutChildren, 'item' );
+
+	return itemsByItemProp;
```

Comment:
All these names feel a bit confusing to me.
E.g. `itemsWithoutChildren` I'd expect it's an array of those items that don't have children (as opposed to items that do have children).

It's also not great performance-wise to loop through the same array over and over again.

What about replacing the whole thing with just a `reduce`?
```jsx
const items = findChildrenWithClosestParent( ... );

return items.reduce( ( acc, { child, parent } ) => {
  acc[ child.props.item ] = omit( {
    ...child.props,
    ...menu: parent?.props?.menu ?? ROOT_MENU,
  }, 'children' );
  return acc;
}, {} );
```

Reply:
Same for `findNavigationMenus` 🙂

Reply:
Yep, that looks a lot cleaner! I went for `cloneDeep` and some mutation though 🤔 I realized we are only doing a shallow copy with spread operator so using `cloneDeep` should be more safe.

Used `delete` to avoid another object copy made by omit.

Looks a lot more readable, but I'm looking forward to your opinion 😄

Reply:
It's probably not a big deal either way.
Let's go with the `cloneDeep` for now, and see if anyone complains about it. 😅